allennlp/data/data_loaders/multitask_data_loader.py

```diff
@@ -206,56 +173,27 @@ def __init__(
             # This will raise a TypeError if any of the underlying loaders doesn't have a length,
-            # which is actually what we want.  If the loader has a length, we set batch_size = 1, so
-            # this will give us the right number of instances.
-            total_instances += self._batch_size_multiplier.get(key, 1.0) * len(loader)
-        if self._drop_last or total_instances % self._batch_size == 0:
-            return int(total_instances) // self._batch_size
+            # which is actually what we want.
+            return self.scheduler.count_batches(
+                {dataset: len(loader) for dataset, loader in self._loaders.items()}
+            )
         else:
-            return int(1 + total_instances) // self._batch_size
+            return self.scheduler.count_batches(
+                {dataset: self._instances_per_epoch for dataset in self._loaders.keys()}
+            )
 
     @overrides
     def __iter__(self) -> Iterator[TensorDict]:
-        # Basic outline: first we _sample_ the instances that we're going to be using for this
-        # epoch, which relies on the scheduler if `self._instances_per_epoch` is not None.  This is
-        # basically just saying how many instances we should use this epoch for each dataset, and we
-        # grab bounded-length iterators over that many instances for each dataset.  Second, we
-        # _schedule_ the epoch's instances into a single list, again relying on the scheduler.
-        # Finally, we take that combined list and yield `batch_size` batches from it.
         epoch_instances = self._get_instances_for_epoch()
-        scheduled_instances = self.scheduler.order_epoch_instances(epoch_instances)
-        batch_instances: List[Instance] = []
-        current_batch_size = 0.0
-        for dataset, instance in scheduled_instances:
-            current_batch_size += self._batch_size_multiplier.get(dataset, 1.0)
-            if current_batch_size > self._batch_size:
-                batch = Batch(batch_instances)
-                tensor_dict = batch.as_tensor_dict()
-                if self.cuda_device is not None:
-                    tensor_dict == nn_util.move_to_device(tensor_dict, self.cuda_device)
-                yield tensor_dict
-                batch_instances = [instance]
-                current_batch_size = self._batch_size_multiplier.get(dataset, 1.0)
-            else:
-                batch_instances.append(instance)
-
-        # Based on how we yield batches above, we are guaranteed to always have leftover instances,
-        # so we don't need a check for that here.
-        if not self._drop_last or current_batch_size == self._batch_size:
-            batch = Batch(batch_instances)
-            tensor_dict = batch.as_tensor_dict()
-            if self.cuda_device is not None:
-                tensor_dict == nn_util.move_to_device(tensor_dict, self.cuda_device)
-            yield tensor_dict
+        return (
+            nn_util.move_to_device(
```

Comment:
Calling `move_to_device` here isn't necessary as long as we pass the target device through to `self._loaders` in `set_target_device()`.

Reply:
That sounds correct. I will check.

Reply:
Oh, never mind. This is definitely necessary.

Reply:
No, wait. We read `Instance`s from the child loaders, not batches. Instances don't have devices. This is the place where we make batches, and only those have devices, so this is where we have to set the device.